editor/connections_dialog.cpp

```diff
@@ -872,19 +872,18 @@ void ConnectionsDock::_make_or_edit_connection() {
 	Ref<Script> scr = target->get_script();
 	if (!scr.is_null() && !ClassDB::has_method(target->get_class(), cd.method)) {
 		// There is a chance that the method is inherited from another script.
-		bool found_inherited_function = false;
-		Ref<Script> inherited_scr = scr->get_base_script();
+		bool found_member_function = false;
+		Ref<Script> inherited_scr = scr;
 		while (!inherited_scr.is_null()) {
-			int line = inherited_scr->get_language()->find_function(cd.method, inherited_scr->get_source_code());
```

Comment:
Someone informed about the original choice to use this function would have to give their opinion on if this should be replaced, it *does* fix the issue for C# at the moment but that's mainly because `find_function` is currently unimplemented for `CSharpScript`

Otherwise it makes sense to me

Reply:
I think `has_method` may return `true` for inherited methods as well. While the old method seems to work off of the source code of the specific file. As such, I don't think this is the right solution. `find_function` needs to be implemented for C# instead. And `make_function` too.

Reply:
Here's the GDScript function for comparison, it only considers the current script:
https://github.com/godotengine/godot/blob/329652b6feac2114c4e30062f18ece36d4e3addc/modules/gdscript/gdscript_editor.cpp#L193-L216

And here's `make_function`:
https://github.com/godotengine/godot/blob/329652b6feac2114c4e30062f18ece36d4e3addc/modules/gdscript/gdscript_editor.cpp#L465-L490

Reply:
Ok, then we need to use something like Roslyn to parse the C# code. The syntax of C# is way more complex than GDScript.

The definition of a partial class can be splitted into multiple files. Unfortunately all nodes declared in C# are partial classes. `find_function` only returns a line number, seems insufficient when there is multiple files for a class.

There is also partial method in C#. A method may be declared at multiple locations in multiple files. `find_function` returns only one value.

So I personally prefer more general method like `has_method`. Scripting language can be very complex, and the editor should not border with actual syntax.

`make_function` has already been implemented for C#.
https://github.com/godotengine/godot/blob/329652b6feac2114c4e30062f18ece36d4e3addc/modules/mono/csharp_script.cpp#L509-L526

 The problem is not how to generate the function signature, but where to insert the code snippet. The insertion logic is hard-coded here.
https://github.com/godotengine/godot/blob/329652b6feac2114c4e30062f18ece36d4e3addc/editor/plugins/script_text_editor.cpp#L309-L325

It always append the code to the end of the file if `find_function` returns -1.

Reply:
> I think `has_method` may return `true` for inherited methods as well.

It seems like `Script::has_method` only checks the methods defined in the script (not inherited). `ScriptInstance::has_method` does include inherited methods, but that's not what is being used here.

However, why would checking for inherited methods be an issue? If we're checking if the method exists to determine whether to add the generated method to the script file, then the existence of a method in a base class should also prevent adding it.

Reply:
@raulsntos The behaviour of Script::has_method is not consistent in GDScript and C#. My another PR is try to make them behave the same https://github.com/godotengine/godot/pull/77273

Reply:
> So I personally prefer more general method like has_method. Scripting language can be very complex, and the editor should not border with actual syntax.

Well, it's not the editor which is handling it, but ScriptLanguage. So it's up to the script language to decide how to handle the case, we just call into its API.

My point is not so much about using specific methods but about the intent of the code. Of course we probably don't want to parse C# intimately for something like this, though since we are doing source code manipulation we just might. A language server would be useful here.

Reply:
> Currently it would add the method handler to `B` which in GDScript means it overrides the method from the parent class. Can we really assume this is the behavior that the user wants?

I think this is a more predictable behavior at least. When you extend some class you don't necessarily know what its composition is. So when you only have your own script before you and try to use the editor GUI to automagically connect to a signal, silently failing to do this because some parent script may have a method with the same name is the worst that we could do.

This entire check operates under the assumption that the workflow is simple: you have a script and you have already created the method and, likely by mistake, try to create it again. So we just skip the step and consider it to be good as is. Which is... eeeeeh, but fine. But I don't think that the same logic can be applied to a more complex case with inherited methods. To me that's definitely a problem that our user must be informed about clearly, since they may not realize why nothing happens or where the conflict comes from.

Reply:
> Using Roslyn sounds like the best solution here, I discussed this with @neikeq when I was working on implementing these methods myself but we weren't happy about the size that the Roslyn libraries add to Godot's assemblies.

I found the method location info is already available during C# source code generation. We can implement `find_function` without introduce new depedency, just dump more metadata to `MethodInfo`.

Reply:
> I suggested a language server though, but I don't know if one exists for C#.

C# users are used to their powerful external IDE. I don't think it is good to launch a LSP and indexing the codebase again for Godot, only for some simple automatic code snippet generation. (Also I wondering how many C# users code in the embeded code editor)

And I suggest the auto code generation to be turned off for languages other than GDScript, it will be much less annoying than the current state. There can be an option to turn auto code snippet insertion off.

Reply:
> So when you only have your own script before you and try to use the editor GUI to automagically connect to a signal, silently failing to do this because some parent script may have a method with the same name is the worst that we could do.

When you put it that way I have agree with you.

> To me that's definitely a problem that our user must be informed about clearly, since they may not realize why nothing happens or where the conflict comes from.

I agree. I think it's important to keep in mind though that in some cases it's not just silent unexpected behavior. As explained earlier, if C# did the same thing as GDScript here, it would result in invalid syntax and it won't compile.

> I'd be fine with using `has_method` and printing popping an error if there is a conflict, suggest using a different name, and let the user figure out the next step.

That makes sense to me. I just want to prevent generating invalid syntax. But I'm worried that this logic is specific to each ScriptLanguage and so it should be moved there.

> I suggested a language server though, but I don't know if one exists for C#.

Roslyn really is the best option here. It's Microsoft's .NET compiler libraries so it allows us to parse C# correctly (same as running `dotnet build`).